package Thread.ThreadCase;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.FutureTask;

public class Test3 {
    /*
        多线程的第三种实现方式 ：
            特点 ： 可以获取到多线程运行的结果

            1. 创建一个MyCallable实现Callable接口
            2. 重写call（是有返回值的， 表示多线程运行的结果）

            3. 创建MyCallable的对象 （表示多线程要执行的任务）
            4. 创建FutureTask对象 （作用管理多线程运行的结果）
            5. 创建Thread对象， 并启动（表示线程）
     */

    static void main() throws ExecutionException, InterruptedException {

        //创建MyCallable的对象 （表示多线程要执行的任务）
        MyCallable3 mc = new MyCallable3();

        //创建FutureTask对象 （作用管理多线程运行的结果）
        FutureTask<Integer> ft = new FutureTask<>(mc);

        //创建线程对象
        Thread t1 = new Thread(ft);

        //启动线程
        t1.start();

        //获取多线程运行结果
        Integer res = ft.get();
        System.out.println(res);

    }
}
